### PR TITLE
Refactor Python API docs to better explain IO binding scenarios

### PR DESCRIPTION
**Pull request by natke**
**Description**: Improve IO Binding docs

**Motivation and Context**
This [thread](https://www.reddit.com/r/MachineLearning/comments/sobfvm/p_what_we_learned_by_accelerating_by_5x_hugging/) highlighted a usage of IO Binding that we had not documented. This change add documentation for the scenario, as well as restructuring the page for better readability.

Changes staged here: https://natke.github.io/onnxruntime/docs/api/python/api_summary.html

### CODE REVIEW

File: docs/python/inference/api_summary.rst [hunk 1]
```diff
@@ -1,60 +1,94 @@
...
+
+	session = onnxruntime.InferenceSession('model.onnx')
+
+	outputs = session.run([output names], inputs)
+
+ONNX and ORT format models consist of a graph of computations, modeled as operators,
+and implemented as optimized operator kernels for different hardware targets.
+ONNX Runtime orchestrates the execution of operator kernels via `execution providers`.
+An execution provider contains the set of kernels for a specific execution target (CPU, GPU, IoT etc).
+Without any configuration, the inference session runs on CPU. Other execution provides are
```
Comment (xadupre) [inline docs/python/inference/api_summary.rst:32]: In case there are multiple providers, onnxruntime prioritizes the first one, then the seconde one...

File: docs/python/inference/api_summary.rst [hunk 2]
```diff
@@ -1,60 +1,94 @@
...
+configured using the `providers` parameter.
+
+.. code-block:: python
+
+	session = onnxruntime.InferenceSession(model, providers=['CUDAExecutionProvider', 'CPUExecutionProvider'])
```
Comment (xadupre) [inline docs/python/inference/api_summary.rst:37]: I would use multiple lines to avoid readers to scroll to see the end of the line.

```
session = onnxruntime.InferenceSession(
       model,
       providers=['CUDAExecutionProvider', 'CPUExecutionProvider'])
```

**Review by cassiebreviu — APPROVED** [review]

File: docs/python/inference/api_summary.rst [hunk 3]
```diff
@@ -133,33 +165,46 @@ Users can bind *OrtValue* (s) directly.
...
+        name='X',
+        device_type='cuda',
+        device_id=0,
+        element_type=np.float32,
+        shape=tuple(x_tensor.shape),
+        buffer_ptr=x_tensor.data_ptr(),
+        )
+
+    ## Allocate the PyTorch tensor for the model output
+	Y_shape = ... # You need to specify the output PyTorch tensor shape
```
Comment (xadupre) [inline docs/python/inference/api_summary.rst:189]: It seems there is one tab `\t` instead of spaces here.

**Review by xadupre — APPROVED** [review]

File: docs/python/inference/api_summary.rst [hunk 4]
```diff
@@ -67,8 +104,6 @@ use IOBinding to put input on CUDA as the follows.
 	session.run_with_iobinding(io_binding)
 	Y = io_binding.copy_outputs_to_cpu()[0]
 
-Scenario 2:
```
Comment (pranavsharma) [inline docs/python/inference/api_summary.rst:70]: I would prefer to mention these sections as scenarios because that's what they really represent - different use cases and how IOBinding can be used in those cases. I found the deletion of Scenario a bit confusing. May be we can call it a different name like Use Case?

File: docs/python/inference/api_summary.rst [hunk 5]
```diff
@@ -133,33 +165,46 @@ Users can bind *OrtValue* (s) directly.
...
-used in the documentation. These don't need to be downloaded
-as they are installed with the package.
-
-.. autofunction:: onnxruntime.datasets.get_example
+.. code-block:: python
 
-Load and run a model
-====================
+    # X is a PyTorch tensor on device
+    session = onnxruntime.InferenceSession('model.onnx')
```
Comment (pranavsharma) [inline docs/python/inference/api_summary.rst:174]: For all the examples under "Data on device" I would've expected the InferenceSession to be constructed using the 'cuda' argument. The current code will not place any operators on CUDA, so the mention of a non-CPU device becomes moot. But I guess this is not introduced in this PR. cc @hariharans29

Comment (jywu-msft) [inline docs/python/inference/api_summary.rst:174]: please change to session = onnxruntime.InferenceSession('model.onnx', providers=['CUDAExecutionProvider'])
i think you want to inference on the gpu?
the input and output is on cuda device

We need to reinforce that providers option is required for non default-CPU providers.

File: docs/python/inference/api_summary.rst (hunk 4 above)
Comment (natke) [reply docs/python/inference/api_summary.rst:70]: I don't have a problem with the word scenario - it was more that I felt like the numbering didn't add anything. If we are breaking  this functionality into a number of canonical scenarios then it would be helpful to name them.

File: docs/python/inference/api_summary.rst (hunk 5 above)
Comment (natke) [reply docs/python/inference/api_summary.rst:174]: Yes, thanks. I've been running into this in another PR I am doing. Can you summarize what happens when:
- you don't specify any providers and you are running on CPU (assume this all works?)
- you don't specify any providers and you are running on GPU (assume this errors?)
- you specify CUDA and you are running on CUDA (assume this all works?)

What would happen before this change?

Thanks @jywu-msft!

Comment (natke) [reply docs/python/inference/api_summary.rst:174]: >  InferenceSession to be constructed using the 'cuda' argument.

What is the 'cuda' argument to InferenceSession? Can you please summarize here? @pranavsharma

File: docs/python/inference/api_summary.rst [hunk 6]
```diff
@@ -133,33 +162,46 @@ Users can bind *OrtValue* (s) directly.
...
+        device_type='cuda',
+        device_id=0,
+        element_type=np.float32,
+        shape=tuple(x_tensor.shape),
+        buffer_ptr=x_tensor.data_ptr(),
+        )
+
+    ## Allocate the PyTorch tensor for the model output
+	Y_shape = ... # You need to specify the output PyTorch tensor shape
+    Y_tensor = torch.empty(Y_shape, dtype=torch.float32, device='cuda:0'.contiguous()
```
[the remark this answers is not in the document]
Comment (natke) [reply docs/python/inference/api_summary.rst]: Fixed

File: docs/python/inference/api_summary.rst (hunk 1 above)
Comment (natke) [reply docs/python/inference/api_summary.rst]: Clarified

File: docs/python/inference/api_summary.rst (hunk 2 above)
Comment (natke) [reply docs/python/inference/api_summary.rst]: Fixed

File: docs/python/inference/api_summary.rst (hunk 5 above)
Comment (jywu-msft) [reply docs/python/inference/api_summary.rst:174]: if the ORT build supports CUDA EP (i.e. it was built with CUDA EP enabled), the providers argument to InferenceSession() is required. It will return error if you don't specify providers. (even if you want to run on CPU)
The only time you can create an InferenceSession() without specifying the providers is you are using a CPU only build. in that scenario, the default CPU provider is implied.

Comment (jywu-msft) [reply docs/python/inference/api_summary.rst:174]: onnxruntime.InferenceSession('model.onnx', providers=['CUDAExecutionProvider'])

Comment (pranavsharma) [reply docs/python/inference/api_summary.rst:174]: I meant specifying the cuda provider in the argument. This is exactly what George is saying as well below.

File: docs/python/inference/api_summary.rst (hunk 4 above)
Comment (pranavsharma) [reply docs/python/inference/api_summary.rst:70]: Sure, I guess it's easier to number them and describe the scenario than coming up with good headings/titles. The main idea is to show different scenarios and how to use IO binding in those scenarios.

File: docs/python/inference/api_summary.rst (hunk 5 above)
Comment (natke) [reply docs/python/inference/api_summary.rst:174]: Great, thank you for explaining that. I will update the docs accordingly

Comment (hariharans29) [reply docs/python/inference/api_summary.rst:174]: @pranavsharma - I think that is because these docs were crafted initially when the provider list was constructed based on build flavor and the user didn't necessarily have to provide the provider list.

Comment (natke) [reply docs/python/inference/api_summary.rst]: Added providers to each of the arguments.

Comment (natke) [reply docs/python/inference/api_summary.rst]: Updated

**Review by pranavsharma — APPROVED** [review]
